src/wmtk/operations/composite/TetEdgeSwap.cpp

```diff
@@ -0,0 +1,150 @@
+#include "TetEdgeSwap.hpp"
+#include <wmtk/simplex/closed_star.hpp>
+#include <wmtk/simplex/open_star.hpp>
+#include <wmtk/simplex/utils/SimplexComparisons.hpp>
+
+
+#include <wmtk/Mesh.hpp>
+
+namespace wmtk::operations::composite {
+TetEdgeSwap::TetEdgeSwap(Mesh& m, long collapse_index)
+    : Operation(m)
+    , m_split(m)
+    , m_collapse(m)
+    , m_collapse_index(collapse_index)
+{}
+
+std::vector<Simplex> TetEdgeSwap::execute(const Simplex& simplex)
+{
+    const auto split_simplicies = m_split(simplex);
+    if (split_simplicies.empty()) return {};
+    assert(split_simplicies.size() == 1);
+
+    // after split
+    const Tuple& split_ret = split_simplicies.front().tuple();
+    assert(!mesh().is_boundary(split_ret));
+    auto iter_tuple = split_ret;
+    std::vector<Tuple> candidate_edge_tuples;
+    while (true) {
+        candidate_edge_tuples.push_back(mesh().switch_edge(iter_tuple));
+        iter_tuple = mesh().switch_tetrahedron(mesh().switch_face(iter_tuple));
+        if (iter_tuple == split_ret) {
+            break;
+        }
+    }
+
+    assert(m_collapse_index < candidate_edge_tuples.size());
+
+    const Tuple& collapse_tuple = candidate_edge_tuples[m_collapse_index];
+
+
+    // before collapse, compute the edges generated by split - edges deleted by collapse
+    auto edges_generated_by_split = m_split.new_edge_tuples();
+
+    const auto v_open_star = simplex::open_star(mesh(), Simplex::vertex(collapse_tuple));
+    const auto e_closed_star = simplex::closed_star(mesh(), Simplex::edge(collapse_tuple));
+    const auto sc = simplex::SimplexCollection::get_intersection(v_open_star, e_closed_star);
+    std::vector<Tuple> edge_tuples_generated_by_swap;
+    std::vector<Tuple> face_tuples_generated_by_swap;
+
+
+    std::array<std::vector<Simplex>, 4> simplices_deleted_by_collapse;
+
+    for (const Simplex& s : sc.simplex_vector()) {
+        simplices_deleted_by_collapse[get_primitive_type_id(s.primitive_type())].emplace_back(s);
+    }
+
+    // check if #edge_generated_by_split > #edges_deleted_by_collapse, if not, return face instead
+    assert(edges_generated_by_split.size() >= simplices_deleted_by_collapse[1].size());
+
+    bool able_to_return_edges = true;
+    if (edges_generated_by_split.size() == simplices_deleted_by_collapse[1].size())
+        able_to_return_edges = false;
+
+    if (able_to_return_edges) {
+        // can use std::difference with sorted vector, worth?
+        for (const auto& e_split : edges_generated_by_split) {
+            bool preserve_flag = true;
+            for (const auto& e_collapse : simplices_deleted_by_collapse[1]) {
```

Comment:
i think for both the edge and face cases daniel's simplex collection code has the ability to find / it sorts things for you